# HG changeset patch
# User Konstantinos Poulios <logari81@googlemail.com>
# Date 1296955468 -3600
#      Sun Feb 06 02:24:28 2011 +0100
# Node ID f6763ab98609fc35886d31e21930d03719d26cee
# Parent  e1da28034c73f5f6e0475c0cc87ab9b1bf34e62f
Prevent race condition in update_axes_layout

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,14 @@
+2011-02-06  Konstantinos Poulios  <logari81@googlemail.com>
+
+	* graphics.cc (updating_axes_layout): New file-scope variable.
+	(axes::properties::update_axes_layout):	Return immediately if
+	updating_axes_layout is true.
+
 2011-02-03  John W. Eaton  <jwe@octave.org>
 
 	* octave.cc: Include <sys/types.h> and <unistd.h> for isatty.
 	Bug #32336.
 
 2011-02-03  Konstantinos Poulios  <logari81@googlemail.com>
 
 	* gl-render.h: (opengl_renderer::draw_axes_titles): Remove prototype.
diff --git a/src/graphics.cc b/src/graphics.cc
--- a/src/graphics.cc
+++ b/src/graphics.cc
@@ -3894,19 +3894,24 @@ axes::properties::update_camera (void)
   // => x_gl_mat2)
   x_gl_mat1 = x_view;
   scale (x_gl_mat1, xd/(xlimits(1)-xlimits(0)), yd/(ylimits(1)-ylimits(0)),
          zd/(zlimits(1)-zlimits(0)));
   translate (x_gl_mat1, -xo, -yo, -zo);
   x_gl_mat2 = x_viewport * x_projection;
 }
 
+static bool updating_axes_layout = false;
+
 void
 axes::properties::update_axes_layout (void)
 {
+  if (updating_axes_layout)
+    return;
+
   graphics_xform xform = get_transform ();
 
   double xd = (xdir_is ("normal") ? 1 : -1);
   double yd = (ydir_is ("normal") ? 1 : -1);
   double zd = (zdir_is ("normal") ? 1 : -1);
 
   const Matrix xlims = xform.xscale (get_xlim ().matrix_value ());
   const Matrix ylims = xform.yscale (get_ylim ().matrix_value ());
@@ -3986,16 +3991,20 @@ axes::properties::update_axes_layout (vo
       zPlane = (dir(0) > 0 ? z_min : z_max);
     else
       zPlane = (dir(1) < 0 ? z_min : z_max);
   else
     zPlane = (dir(2) < 0 ? z_min : z_max);
   zPlaneN = (zPlane == z_min ? z_max : z_min);
   fz = (z_max-z_min)/sqrt(dir(0)*dir(0)+dir(1)*dir(1));
 
+  unwind_protect frame;
+  frame.protect_var (updating_axes_layout);
+  updating_axes_layout = true;
+
   update_ticklengths ();
 
   xySym = (xd*yd*(xPlane-xPlaneN)*(yPlane-yPlaneN) > 0);
   zSign = (zd*(zPlane-zPlaneN) <= 0);
   xyzSym = zSign ? xySym : !xySym;
   xpTick = (zSign ? xPlaneN : xPlane);
   ypTick = (zSign ? yPlaneN : yPlane);
   zpTick = (zSign ? zPlane : zPlaneN);
